File: app/src/main/java/com/dumuzeyn/mp3player/PlaybackUiState.java
package com.dumuzeyn.mp3player;

import java.util.ArrayList;

/** UI projection of Media3 state. Media3 remains the playback source of truth. */
final class PlaybackUiState {
    final ArrayList<Track> queue = new ArrayList<>();
    private PlaybackSnapshot snapshot = PlaybackSnapshot.empty();
    private String cachedMediaId = "";
    private int cachedTrackIndex = -1;
    long sleepTimerEndsAt;

    PlaybackSnapshot snapshot() {
        return snapshot;
    }

    void updateSnapshot(PlaybackSnapshot value) {
        PlaybackSnapshot next = value == null ? PlaybackSnapshot.empty() : value;
        if (!snapshot.currentMediaId.equals(next.currentMediaId)) {
            cachedMediaId = "";
            cachedTrackIndex = -1;
        }
        snapshot = next;
    }

    int currentTrackIndex(LibraryState library) {
        if (snapshot.currentMediaId.isEmpty()) {
            return -1;
        }
        if (cachedMediaId.equals(snapshot.currentMediaId)
                && cachedTrackIndex >= 0 && cachedTrackIndex < library.tracks.size()
                && MediaItemMapper.matchesMediaId(
                        library.tracks.get(cachedTrackIndex), snapshot.currentMediaId)) {
            return cachedTrackIndex;
        }
        for (int index = 0; index < library.tracks.size(); index++) {
            if (MediaItemMapper.matchesMediaId(
                    library.tracks.get(index), snapshot.currentMediaId)) {
                cachedMediaId = snapshot.currentMediaId;
                cachedTrackIndex = index;
                return index;
            }
        }
        cachedMediaId = snapshot.currentMediaId;
        cachedTrackIndex = -1;
        return -1;
    }

    boolean isPlaying() {
        return snapshot.playWhenReady
                && snapshot.phase != PlaybackPhase.ENDED
                && snapshot.phase != PlaybackPhase.ERROR;
    }

    int repeatMode() {
        return RepeatModeMapper.fromMedia3(snapshot.repeatMode);
    }

    boolean shuffleEnabled() {
        return snapshot.shuffleEnabled;
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/PlaybackUiState.kt
package com.dumuzeyn.mp3player

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/** Lifecycle-independent UI projection of the authoritative Media3 state. */
class PlaybackUiState {
    @JvmField val queue = ArrayList<Track>()
    @JvmField var sleepTimerEndsAt: Long = 0L

    private val mutableState = MutableStateFlow(PlaybackSnapshot.empty())
    private var cachedMediaId = ""
    private var cachedTrackIndex = -1

    fun state(): StateFlow<PlaybackSnapshot> = mutableState.asStateFlow()

    fun snapshot(): PlaybackSnapshot = mutableState.value

    fun updateSnapshot(value: PlaybackSnapshot?) {
        val next = value ?: PlaybackSnapshot.empty()
        if (mutableState.value.currentMediaId != next.currentMediaId) {
            cachedMediaId = ""
            cachedTrackIndex = -1
        }
        mutableState.value = next
    }

    fun currentTrackIndex(library: LibraryState): Int {
        val currentMediaId = mutableState.value.currentMediaId
        if (currentMediaId.isEmpty()) return -1
        if (
            cachedMediaId == currentMediaId &&
            cachedTrackIndex in library.tracks.indices &&
            MediaItemMapper.matchesMediaId(library.tracks[cachedTrackIndex], currentMediaId)
        ) {
            return cachedTrackIndex
        }
        val index = library.tracks.indexOfFirst {
            MediaItemMapper.matchesMediaId(it, currentMediaId)
        }
        cachedMediaId = currentMediaId
        cachedTrackIndex = index
        return index
    }

    fun isPlaying(): Boolean {
        val snapshot = mutableState.value
        return snapshot.playWhenReady &&
            snapshot.phase != PlaybackPhase.ENDED &&
            snapshot.phase != PlaybackPhase.ERROR
    }

    fun repeatMode(): Int = RepeatModeMapper.fromMedia3(mutableState.value.repeatMode)

    fun shuffleEnabled(): Boolean = mutableState.value.shuffleEnabled
}
